### PR TITLE
Box all the things

### PR DESCRIPTION
**Pull request by mweststrate**
First class boxes, references, etc. Should address irregularities related to references, add support for references inside `maybe`, `map`, `array` etc

### CODE REVIEW

**Comment by robinfehr** [comment]
awesome - reached that problem just now :)

**Comment by mattiamanzati** [comment]
Possible enhancement: Since typecheck() is recursive, we can remove all the typechecks from instantiate and assume that snapshot is a valid snapshot. Maybe typecheck should be performed only in create() and applySnapshot()

File: src/core/node.ts [hunk 1]
```diff
@@ -11,6 +11,7 @@ export class Node  {
     // optimization: these fields make MST memory expensive for primitives. Most can be initialized lazily, or with EMPTY_ARRAY on prototype
     readonly nodeId = ++nextNodeId
     readonly type: IType<any, any>
```
Comment (mweststrate) [inline src/core/node.ts]: maybe rename type -> actualType to make it more clear?

File: src/types/complex-types/map.ts [hunk 2]
```diff
@@ -42,6 +42,7 @@ export class MapType<S, T> extends ComplexType<{[key: string]: S}, IExtendedObse
     createNewInstance() {
         // const identifierAttr = getIdentifierAttribute(this.subType)
         const map = observable.shallowMap()
+        extras.getAdministration(map).dehancer = unbox
```
Comment (mweststrate) [inline src/types/complex-types/map.ts]: dehancer is the worst name, looking for a better alternative :-P

Comment (mattiamanzati) [inline src/types/complex-types/map.ts]: boxer and unboxer? 👖

**Pull request merged**